aten/src/ATen/native/quantized/cpu/qnnpack_utils.h

```diff
@@ -84,3 +89,58 @@ inline std::pair<uint8_t, uint8_t> activationLimits(
   }
 }
 #endif
+
+namespace {
+std::vector<float> generate_requantization_scales(
+    const at::Tensor& weight_scales,
+    const float input_scale,
+    const float output_scale) {
+  auto num_output_channels = weight_scales.numel();
+  float* weight_scales_data = weight_scales.data_ptr<float>();
+  std::vector<float> requant_scales(num_output_channels, 1.f);
+  for (int i = 0; i < num_output_channels; ++i) {
+    requant_scales[i] = weight_scales_data[i] * input_scale / output_scale;
+  }
+  return requant_scales;
+}
+
+std::pair<at::Tensor, at::Tensor> make_zero_points_and_scales_tensor(
+    const at::Tensor& weight_contig
+    ) {
+  auto num_output_channels = weight_contig.size(0);
+  const auto qtype = weight_contig.qscheme();
+  at::Tensor weight_zp =
+    at::native::empty_cpu({1}, at::device(at::kCPU).dtype(at::kQUInt8));
```

Comment:
Why quint8? fbgemm prepack step seems to be using int32_t.
Also why call empty_cpu explicitly here? Wouldn't empty() work?

Reply:
Because qnnpack uses uint8 to store zero points. For empty_cpu it is to by pass all the dispatch overhead.

Reply:
Is there a lot of overhead? Reason I ask is that we are still using aten::empty in other parts of the quantized code which is also run on mobile.

Reply:
Well, in FBNet I saw this as an issue and calling directly the native function was getting rid of some overhead. Now it depends on the network. If the ops are taking much less time than such overhead matters.

Reply:
Actually in this case, it should be much of overhead, because this is done only once.